src/Halogen/Component.purs

```diff
@@ -1,64 +1,214 @@
-hoistComponent :: forall m n req res. (forall a. m a -> n a) -> Component m req res -> Component n req res
-hoistComponent f sf = (f <$>) <$> sf
-
--- | Combine two components into a single component.
--- |
--- | The first argument is a function which combines the two rendered HTML documents into a single document.
--- |
--- | This function works on request and response types by taking the _sum_ in each component. The left summand
--- | gets dispatched to (resp. is generated by) the first component, and the right summand to the second component.
-combine :: forall m req1 req2 res1 res2.
-             (Functor m) =>
-             (forall a. HTML a -> HTML a -> HTML a) ->
-             Component m req1 res1 ->
-             Component m req2 res2 ->
-             Component m (Either req1 req2) (Either res1 res2)
-combine f = mergeWith f1
+import Control.Monad.Aff (Aff())
+import Control.Monad.Eff.Class (liftEff)
+import Control.Monad.Eff.Console (CONSOLE())
+import Control.Monad.Eff.Ref (REF(), Ref(), newRef, writeRef, readRef)
+import Control.Monad.Free (Free(), FreeC(), runFreeM, runFreeCM)
+import Control.Monad.Rec.Class (MonadRec)
+import Control.Monad.State (State(), runState)
+import Control.Monad.State.Class (MonadState, get, gets, put)
+import Control.Monad.State.Trans (StateT(), runStateT)
+import Control.Monad.Trans (MonadTrans, lift)
+import Control.Plus (Plus, empty)
+
+import Data.Bifunctor (lmap)
+import Data.Coyoneda (Natural())
+import Data.DOM.Simple.Types (HTMLElement())
+import Data.Either (Either(..), either)
+import Data.Functor.Coproduct (Coproduct(), coproduct, left, right)
+import Data.Maybe (Maybe(..))
+import Data.Tuple (Tuple(..), fst, snd)
+import Data.Void (Void())
+
+import DOM (DOM())
+
+import Halogen.HTML.Renderer.VirtualDOM (renderHTML)
+import Halogen.HTML.Core (HTML(..), install)
+import Halogen.Internal.VirtualDOM (VTree(), createElement, diff, patch)
+
+import qualified Data.Map as M
+
+newtype Component s f g p = Component
+  { render :: State s (HTML p (f Unit))
+  , query  :: forall i. f i -> StateT s g i
+  }
+
+type ComponentF s f = Component s (Free f)
+type ComponentFC s f = Component s (FreeC f)
+
+type Render s p i = s -> HTML p i
+type RenderF s p f = s -> HTML p (Free f Unit)
+type RenderFC s p f = s -> HTML p (FreeC f Unit)
+
+type Eval f s g = Natural f (StateT s g)
+
+renderComponent :: forall s f g p. Component s f g p -> s -> Tuple (HTML p (f Unit)) s
+renderComponent (Component c) = runState c.render
+
+queryComponent :: forall s f g p i. Component s f g p -> f i -> s -> g (Tuple i s)
+queryComponent (Component c) q = runStateT (c.query q)
+
+instance functorComponent :: Functor (Component s f g) where
+  map f (Component c) =
+    Component { render: lmap f <$> c.render
+              , query: c.query
+              }
+
+component :: forall s f g p. (s -> HTML p (f Unit)) -> (forall i. f i -> StateT s g i) -> Component s f g p
```

Comment:
This one is not using the type aliases like the other ones, which makes it inconsistent and its signature overly complex.